eval/topdown.go

```diff
@@ -489,18 +563,55 @@ func evalRefRec(ctx *TopDownContext, ref opalog.Ref, iter TopDownIterator, path
 
 	headVar, isVar := head.Value.(opalog.Var)
 
-	if !isVar || len(path) == 0 {
+	// Handle head of reference.
+	if isVar && len(path) == 0 {
+
+		// Check if head has a binding. E.g., x = [1,2,3], x[i] = 1
+		binding := ctx.Bindings.Get(headVar)
+		if binding != nil {
+			return evalRefRuleResult(ctx, ref, ref[1:], binding, iter)
+		}
+
 		path = append(path, head)
-		return evalRefRec(ctx, tail, iter, path)
+		node, err := ctx.Store.Lookup(path)
+		if err != nil {
+			switch err := err.(type) {
+			case *StorageError:
+				if err.Code == StorageNotFoundErr {
+					return nil
+				}
+			}
+			return err
+		}
+
+		switch node := node.(type) {
+		case []*opalog.Rule:
+			for _, rule := range node {
+				if err := evalRefRule(ctx, ref, path, rule, iter); err != nil {
+					return err
+				}
+			}
+			return nil
+		default:
+			return evalRefRec(ctx, tail, iter, path)
+		}
 	}
 
-	binding, isBound := ctx.Bindings[headVar]
+	// Handle constants in reference.
+	if !isVar {
+		path = append(path, head)
+		return evalRefRec(ctx, tail, iter, path)
+	}
 
-	if isBound {
+	// Binding already exists for variable. Treat it as a constant.
+	binding := ctx.Bindings.Get(headVar)
```

Comment:
At the beginning of the function, we're looking up the path in Storage once ref is empty (which I assume means the root is a doc in Storage).  But here, the head could be a var and the binding could be a dictionary/object/set, right?  So then the root of the doc isn't in storage.  Or am I misreading this?